tests/dags/providers/provider_api_scripts/test_finnish_museums.py

```diff
@@ -30,73 +36,230 @@ def _get_resource_json(json_name):
+            ("2020-04-01T03:45:00Z", "2020-04-01T03:50:00Z"),
+            ("2020-04-01T03:50:00Z", "2020-04-01T03:55:00Z"),
+            ("2020-04-01T03:55:00Z", "2020-04-01T04:00:00Z"),
+        ]
+
+        actual_pairs_list = fm._get_timestamp_pairs("test_building")
+        # Formatting the timestamps so the test will be more readable
+        formatted_actual_pairs_list = [
+            (fm.format_ts(x), fm.format_ts(y)) for x, y in actual_pairs_list
+        ]
+        assert formatted_actual_pairs_list == expected_pairs_list
+
+
+def test_ingest_records_halts_early_if_the_total_count_has_been_exceeded():
+    # TODO: Clean this test up, better doc strings
+    # Top level doc string which describes what is expected to happen over the course
+    # of the test.
+    fm = FinnishMuseumsDataIngester(date=FROZEN_DATE)
+    fm.buildings = ["test_building"]
+
+    # Mock _get_timestamp_pairs to return two timestamps. Later we'll
+    # mock the first interval to simulate the bug, and the second will
+    # work as normal.
+    with patch.object(fm, "_get_timestamp_pairs") as ts_mock:
+        ts_mock.return_value = [
+            (
+                datetime(2020, 4, 1, 11, 0, tzinfo=timezone.utc),
+                datetime(2020, 4, 1, 12, 0, tzinfo=timezone.utc),
+            ),
+            (
+                datetime(2020, 4, 1, 12, 0, tzinfo=timezone.utc),
+                datetime(2020, 4, 1, 13, 0, tzinfo=timezone.utc),
+            ),
+        ]
+
+        mock_response = {
+            "status": "ok",
+            "records": [MagicMock(), MagicMock()],  # Two mock records,
+            "resultCount": 2,  # Claim there are only two records total
+        }
+
+        with (
+            patch.object(fm, "get_response_json") as get_mock,
+            patch.object(fm, "process_batch", return_value=2) as process_mock,
+            patch("common.slack.send_alert") as slack_alert_mock,
+        ):
+            # mock get_response_json calls in order
+            get_mock.side_effect = [
+                # First call, get the first page which says there are 2 records
+                mock_response,
+                # Second call, have it get a second page even though we've already ingested 2 records
+                mock_response,
+                # Third call, this should be for the SECOND timeslice (should've returned early last time)
+                mock_response,
+                # Fourth call halts ingestion entirely
+                None,
+            ]
+
+            # Now try to ingest records
+            fm.ingest_records()
+
+            # get_mock should have been called 4 times
+            assert get_mock.call_count == 4
+            # process_mock should only have been called twice. The second 'batch' for the first time slice
+            # should have returned early
+            assert process_mock.call_count == 2
+            # Slack should have been alerted to the issue
+            assert slack_alert_mock.called is True
```

Comment:
note to self:  test that get_response_json was called specifically twice for the first timeslice and twice for the second, by checking the query params it was called with.